feat(connlib): utilise GSO for UDP sockets

thomaseizinger opened the pull request:
## Context

At present, `connlib` sends UDP packets one at a time. Sending a packet requires us to make a syscall which is quite expensive. Under load, i.e. during a speedtest, syscalls account for over 50% of our CPU time [0]. In order to improve this situation, we need to somehow make use of GSO (generic segmentation offload). With GSO, we can send multiple packets to the same destination in a single syscall.

The tricky question here is, how can we achieve having multiple UDP packets ready at once so we can send them in a single syscall? Our TUN interface only feeds us packets one at a time and `connlib`'s state machine is single-threaded. Additionally, we currently only have a single `EncryptBuffer` in which the to-be-sent datagram sits.

## 1. Stack-allocating encrypted IP packets

As a first step, we get rid of the single `EncryptBuffer` and instead stack-allocate each encrypted IP packet. Due to our small MTU, these packets are only around 1300 bytes. Stack-allocating that requires a few memcpy's but those are in the single-digit % range in the terms of CPU time performance hit. That is nothing compared to how much time we are spending on UDP syscalls. With the `EncryptBuffer` out the way, we can now "freely" move around the `EncryptedPacket` structs and - technically - we can have multiple of them at the same time.

## 2. Implementing GSO

The GSO interface allows you to pass multiple packets **of the same length and for the same destination** in a single syscall, meaning we cannot just batch-up arbitrary UDP packets. Counterintuitively, making use of GSO requires us to do more copying: In particular, we change the interface of `Io` such that "sending" a packet performs essentially a lookup of a `BytesMut`-buffer by destination and packet length and appends the payload to that packet.

## 3. Batch-read IP packets

In order to actually perform GSO, we need to process more than a single IP packet in one event-loop tick. We achieve this by batch-reading up to 50 IP packets from the mpsc-channel that connects `connlib`'s main event-loop with the dedicated thread that reads and writes to the TUN device. These reads and writes happen concurrently to `connlib`'s packet processing. Thus, it is likely that by the time `connlib` is ready to process another IP packet, multiple have been read from the device and are sitting in the channel. Batch-processing these IP packets means that the buffers in our `GsoQueue` are more likely to contain more than a single datagram.

Imagine you are running a file upload. The OS will send many packets to the same destination IP and likely max MTU to the TUN device. It is likely, that we read 10-20 of these packets in one batch (i.e. within a single "tick" of the event-loop). All packets will be appended to the same buffer in the `GsoQueue` and on the next event-loop tick, they will all be flushed out in a single syscall.

## Results

Overall, this results in a significant reduction of syscalls for sending UDP message. In [1], we spend only a total of 16% of our CPU time in `udpv6_sendmsg` whereas in [0] (main), we spent a total of 34%. Do note that these numbers are relative to the total CPU time spent per program run and thus can't be compared directly (i.e. you cannot just do 34 - 16 and say we now spend 18% less time sending UDP packets). Nevertheless, this appears to be a great improvement.

In terms of throughput, we achieve a ~60% improvement in our benchmark suite. That one is running on localhost though so it might not necessarily be reflect like that in a real network.

[0]: https://share.firefox.dev/4hvoPju
[1]: https://share.firefox.dev/4frhCPv

thomaseizinger committed:
Stack-allocate encrypted packets

thomaseizinger committed:
Use GSO to send multiple packets at once

thomaseizinger committed:
Improve `Io` interface

thomaseizinger committed:
Introduce dedicated module

jamilbk commented on rust/connlib/tunnel/src/io.rs:
```diff
@@ -146,26 +149,32 @@ impl Io {
                 let deadline = timeout.deadline().into();
                 self.timeout.as_mut().take(); // Clear the timeout.
 
+                // Piggy back onto the timeout we already have.
```
Would it be helpful to state the periodicity here? E.g. every 60ms?

thomaseizinger replied:
The periodicity here is whatever the state machine requests! Could be as long as a 30s on an idle connection :)

I'll add a more detailed comment.

jamilbk reviewed (dismissed):
Trading a few mem copies for fewer syscalls seems to make sense.

Curious where a bit more time profiling and optimization can get us! It would be nice to be able to have benchmarks / blog post to point users to on sales calls. Sizing is one of the main questions / friction points admins have when deploying Gateways.

thomaseizinger committed:
Activate fast apple datapath for quinn-udp

thomaseizinger committed:
Log num packets and segment size in `wire`

thomaseizinger committed:
Read IP packets in batches of 50

thomaseizinger committed:
Introduce shared `Buffers`

thomaseizinger committed:
Name TUN send-recv thread

thomaseizinger committed:
Add changelog entries

thomaseizinger committed:
Also print trailing bytes

thomaseizinger committed:
Remove unnecessary force-flush

thomaseizinger committed:
Don't check for > 0

This always returns > 0 unless the channel is closed (which it should
never be).

thomaseizinger committed:
Update `quinn-udp`

thomaseizinger committed:
Don't hard code number of segments

thomaseizinger committed:
Respect `max_gso_segments`

thomaseizinger committed:
Don't log trailing bytes

We never have trailing bytes because `GsoQueue` always tracks by
segment-size.

thomaseizinger committed:
Inline `try_send`

thomaseizinger committed:
Log `wire::net::send` after splitting

jamilbk reviewed (approved):
Tests seemed ok on iOS and Windows. Not much if any speed difference.

thomaseizinger committed:
Document re-allocation behaviour
